torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+  if (target_name == nullptr || target_name.str() != "xla_mark_tensor") {
+    return false;
+  }
+  return true;
+}
+
+struct BoundaryMetadata {
+  std::string name;
+  int64_t id;
+  int64_t pos;
+  bool is_input;
+  std::unordered_map<std::string, json> attrs;
+
+  std::string boundary_id() const { return absl::StrCat(name, "__", id); }
+
+  auto uid() const { return std::forward_as_tuple(name, id, pos, is_input); }
+
+  bool operator==(const BoundaryMetadata& other) const {
+    return uid() == other.uid();
+  }
+  bool operator<(const BoundaryMetadata& other) const {
+    return uid() < other.uid();
+  }
+
+  static std::unique_ptr<BoundaryMetadata> Parse(llvm::StringRef str) {
+    auto j = json::parse(str, /*cb=*/nullptr, /*allow_exceptions=*/false);
+    return Build(j);
+  }
+
+ private:
+  template <typename T>
+  static bool CopyJsonValue(const nlohmann::basic_json<>& j,
+                            llvm::StringRef key, json::value_t expected_type,
+                            T& to) {
+    auto kv = j.find(key);
+
+    if (kv == j.end()) {
+      return false;
+    }
+    if (kv.value().type() != expected_type) {
+      return false;
+    }
+    kv.value().get_to(to);
+    return true;
+  }
+
+  static std::unique_ptr<BoundaryMetadata> Build(
+      const nlohmann::basic_json<>& j) {
+    BoundaryMetadata metadata;
+
+    bool is_valid_metadata_json =
+        CopyJsonValue(j, "name", json::value_t::string, metadata.name) &&
+        CopyJsonValue(j, "id", json::value_t::number_unsigned, metadata.id) &&
+        CopyJsonValue(j, "pos", json::value_t::number_unsigned, metadata.pos) &&
+        CopyJsonValue(j, "is_input", json::value_t::boolean, metadata.is_input);
+
+    if (!is_valid_metadata_json) {
+      return nullptr;
+    }
+
+    if (auto kv = j.find("attr"); kv != j.end() && kv.value().is_object()) {
+      auto& attrs_j = kv.value();
+      for (auto attr_j = attrs_j.begin(); attr_j != attrs_j.end(); ++attr_j) {
+        metadata.attrs.insert({attr_j.key(), attr_j.value()});
+      }
+    }
+    return std::make_unique<BoundaryMetadata>(std::move(metadata));
+  }
+};
+
+class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
+ public:
+  explicit BuildStableHLOCompositePass()
+      : mlir::OperationPass<mlir::ModuleOp>::OperationPass(
+            mlir::TypeID::get<BuildStableHLOCompositePass>()) {}
+
+  ~BuildStableHLOCompositePass() override = default;
+
+  void runOnOperation() override {
+    mlir::ModuleOp module_op = getOperation();
+    llvm::SmallVector<mlir::func::FuncOp> func_ops(
+        module_op.getOps<mlir::func::FuncOp>());
+    for (mlir::func::FuncOp& func_op : func_ops) {
+      llvm::DenseMap<const mlir::Operation*, size_t> op_line_num =
+          BuildOperationsLineNumberMap(func_op);
+      for (auto op : func_op.getOps<mlir::stablehlo::CustomCallOp>()) {
+        BuildStableHLOCompositeOp(op.getOperation(), op_line_num);
```

Comment:
`if IsXlaMarkTensorOp(op.getOperation)`?

When should this method be called? Only on outputs and only for MarkTensor ops?

Reply:
Can we write this in a way where this method should be expected to succeed and give an error message if it fails.

Currently it fails silently (if it isn't the right custom call target, if JSON parsing fails, etc). Consider making stronger use of `FailureOr<T>` which can return a failure message or a type on success.

Reply:
> When should this method be called? Only on outputs and only for MarkTensor ops?
Yes. There is a check right after `GetBoundaryMetadata` in the function.

> Can we write this in a way where this method should be expected to succeed and give an error message if it fails.
This is intended as the composite is optional and the invalid mark_tensor custom_call can still be removed afterwards. Do you suggest to fail the conversion if the composite is not complete?

Reply:
Personal preference is yes - From a functional perspective, if a boundary is marked, it is likely intentional. If we must export with a boundary, it is probably better to know during export time as opposed to much later in the pipelines, whether it manifests via poor performance on some hardware, failed pattern match later in compilation, etc.

From a usability perspective - If we silently fail, what is the next step to diagnosing an issue? If we cannot honor that boundary, we should at a minimum give the reason why.